internal/testrunner/runners/system/servicedeployer/terraform.go

```diff
@@ -42,6 +46,38 @@ type TerraformServiceDeployer struct {
 	definitionsDir string
 }
 
+// addTerraformOutputs method reads the terraform outputs generated in the json format and
+// adds them to the custom properties of ServiceContext and can be used in the handlebars template
+// like `{{TF_OUTPUT_queue_url}}` where `queue_url` is the output configured
+func addTerraformOutputs(outCtxt ServiceContext) error {
+	// Read the `output.json` file where terraform outputs are generated
+	content, err := os.ReadFile(outCtxt.OutputDir + terraformOutputJsonFile)
+	if err != nil {
+		return fmt.Errorf("failed to read terraform output file: %w", err)
+	}
+
+	// https://github.com/hashicorp/terraform/blob/v1.4.6/internal/command/views/output.go#L217-L222
+	type OutputMeta struct {
+		Value interface{} `json:"value"`
+	}
+
+	// Unmarshall the data into `terraformOutputs`
+	logger.Debug("Unmarshalling terraform output json")
+	var terraformOutputs map[string]OutputMeta
+	if err = json.Unmarshal(content, &terraformOutputs); err != nil {
+		return fmt.Errorf("error during json Unmarshal %w", err)
+	}
+
+	if len(terraformOutputs) > 0 && outCtxt.CustomProperties == nil {
+		outCtxt.CustomProperties = make(map[string]any, len(terraformOutputs))
+	}
+	// Prefix variables names with TF_OUTPUT_
+	for k, outputs := range terraformOutputs {
+		outCtxt.CustomProperties[terraformOutputPrefix+k] = outputs.Value
+	}
```

Comment:
Maybe it can be return early if there are no terraformOutputs

```suggestion
	if len(terraformOutputs) == 0 {
		return nil
	}
	if outCtxt.CustomProperties == nil {
		outCtxt.CustomProperties = make(map[string]any, len(terraformOutputs))
	}
	// Prefix variables names with TF_OUTPUT_
	for k, outputs := range terraformOutputs {
		outCtxt.CustomProperties[terraformOutputPrefix+k] = outputs.Value
	}
```

Reply:
There is not any notable computing that is happening if `terraformOutputs` is empty.

But yeah , May be more modular and cleaner.